### ISSUE TITLE
Bidder logic framework for Cumulus

### ISSUE THREAD

**Issue by rphmeier** [paritytech/cumulus#2154]
We should provide a pluggable framework in Cumulus for collators to determine when to bid on parathread blockspace. Parathread node authors should be able to use this framework to choose on the basis of:
  1. Relay chain blocks since last parachain block
  2. Block reward when authoring a parachain block (some runtimes may exponentially increase the block reward exponentially based on (1) )
  3. Transaction fees available in the parachain
  4. Other user-defined logic

**Comment by eskimor** [comment]
Size of transaction pool

**Comment by bkchr** [comment]
> Size of transaction pool

Don't you get this indirectly by "transaction fees available"? 

In general for this issue, not sure we need to create some kind of framework. I mean we should probably provide some implementations using different kind of strategies that can be used as examples.

**Comment by rphmeier** [comment]
Yeah, I mean the "framework" only loosely. The information just needs to be available, as well as some easy hook to trigger bidding.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/on-demand-parachains/2208/1

**Comment by Muhammad4599** [comment]
**__**

**Comment by burdges** [comment]
I'd forget 3 completely since someone else could pay too.  I'd propose this scheme:

At the relay chain side, we've a `purchase_blocks` call which spends some amount of DOTs on parathread blocks by some minimum larger than the current parathread block price, and provides an opaque value `buyer`.  All it does is grow some `buyers: Vec<(dots,buyer)>` in the parathread's state on the relay chain.  An `accept_buyers` intrinsic/message from the parachian side can empty `buyers` and increase some `bought: i32` value by the number of blocks this number of dots covers.  A parachain can make a block so long as either `buyers.len() + bought > 0`.

Parathreads have completely self defined runtime logic for determining the next block, including tricks like sassafras or whisk.  We want parachains to expose some `allowed_connection` function for validating the next block producer though, which returns some `Resutl<Allowed,NotAllowed>`. 

A good backing statement subtracts 1 from `bought`, but the parablock itself could execute `accept_buyers` so that even if `bought=0` then we expect `bought >= 0` after inclusion.  If this is not true then the parablock is invalid.

If otoh we have `bought < 0` after inclusion then instead a backer could (optionally) make a bad backing statement, which executes the block but does not update the state roots.  A bad backing statement might just lock the parachain, or deduct form a random element in `buyers` or simply call `buyers.clear()`.  A bad backing statement is invalid if `bought >= 0` after execution, slashing the backer.

A parachain is responsible for ensuring that `allowed_connection` never permits backers to make bad backing statements. We should discuss if this is too difficult a rule for when we want `allowed_connection` to be run.

**Comment by eskimor** [comment]
I am not sure I get how this maps to the listed requirements. For forgetting (3): Yes someone else could pay too, but you still need coordination: Who else should pay? When do you decide that you should pay? We need some coordination here, some coordination that takes into account that collators can be offline.

**Comment by eskimor** [comment]
Uh I think I am starting to getting it: You are thinking probabilistic, that orders should stay valid for long periods of time and can be consumed whenever. This is not how on-demand works though. An order placed will be served as soon as possible and then you have to produce a block, otherwise you just wasted a core and money.

**Comment by burdges** [comment]
Yes, I'm thinking orders should stay valid for a long period or time.  It's clear we need restrictions upon opening connections of course, but unclear why this means purchases expire quickly. 

As a clarification..

A parachain's own code should only send the `accept_buyers` message when it also translates those opaque `buyer` fields into its own restrictions upon block production within its own state.  

A parachain's own block production rules plus `accept_connection` should then obey these restriction. 

A prachain could enforce that only the real buyer can make the block, or tax them to open some blocks for everyone, or do probabalistic thing, or whatever.

I'm now dubious this bad backing statement makes sense, maybe `allow_connection` should just always produce a valid payment, but the backer gets more money if the block becomes available, finalized, etc, via the rewards system which we've never quite finished.

**Comment by burdges** [comment]
Anyways, we've always wanted to permit parachains to design their own block production strategies.  We do imho need an interface that excludes some really stupid ones, for which no `allow_connections` function makes sense, like proof-of-work.  We'll save ourselves lots of trouble though if we've one interface broad enough to permit aura, sassafras, whisk, and some others, in conjunction with flexible economics rules.

I think parathreads should be able to impose a tax upon parablock buys, but the relay chain should know nothing about how their tax works.  This is one reason for the opaque buyer field being processed on the parachain side.

At the same time, we need parathreads to make blocks even if they have exhausted their balance, which is why the `accept_buyers` can come inside the block that spends the buyer here, aka we permit `bought=0` both before and after the block.

**Comment by burdges** [comment]
As an alternative..

We could decide that on-deman parachains should've one or two good block production and economic models, but they need not be anywhere near as flexible as full parachains.  If you want your own specialized block production and economic models, then just rent a full parachain slot.  It only costs like $25k per year, so we'd be over engineering if we try to make the on-demand option super flexible. 

Is this closer to how others view the problem?

**Comment by burdges** [comment]
Alright @eskimor and I discussed this, all this pre-scheduling stuff exists to avoid doing a separate protocol for parathreads.  We do more complex code elsewhere but avoid subtle refactoring.
